File: IFAA/originDataScreen.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Mar 25 22:07:38 2022

@author: ss
"""
## Original screen function to select independent reference taxa

import numpy as np
from utility import *
import timeit
import multiprocessing as mp
import math
import joblib
from dataInfo import *
from dataRecovTrans import *
from AIcalcu import *

from sklearn.linear_model import LassoCV

from statsmodels.api import OLS
from statsmodels.stats.multitest import multipletests


def originDataScreen(
  data,
  testCovInd,
  nRef,
  paraJobs,
  refTaxa,
  standardize,
  sequentialRun,
  allFunc,
  Mprefix,
  covsPrefix,
  binPredInd,
  adjust_method,
  seed,
  maxDimensionScr=434*5*10**5):
    
    results = {}
    
    # load data info
    basicInfo=dataInfo(data=data,Mprefix=Mprefix,
                   covsPrefix=covsPrefix,
                   binPredInd=binPredInd)
    
    taxaNames=basicInfo['taxaNames']
    nTaxa=basicInfo['nTaxa']
    nPredics=basicInfo['nPredics']
    rm(basicInfo)
    
    nNorm=nTaxa-1
    nAlphaNoInt=nPredics*nNorm
    nAlphaSelec=nPredics*nTaxa
    
    countOfSelec=np.zeros(nAlphaSelec)
    
    # overwrite nRef if the reference taxon is specified
    nRef=len(refTaxa)
    
    startT1=timeit.default_timer()
    if paraJobs is None:
        availCores = mp.cpu_count()
        if isinstance(availCores, int): 
            paraJobs = max(1, availCores-2)
    
    batch=paraJobs
    forLoopN=math.ceil(nRef/batch)
    
    if not sequentialRun:
        print(paraJobs, " parallel jobs are registered for analyzing ", nRef, " reference taxa in Phase 1")

        
i = 1

def foreachUnitRun(i):
    
    ii=which(taxaNames==refTaxa[i])
    dataForEst=dataRecovTrans(data=data,
                              ref=refTaxa[i],
                              Mprefix=Mprefix,
                              covsPrefix=covsPrefix)

    xTildLongTild_i=dataForEst['xTildalong']
    yTildLongTild_i=dataForEst['UtildaLong']
    rm(dataForEst)
    
    maxSubSamplSiz=np.min( (50000.0, np.floor(maxDimensionScr / xTildLongTild_i.shape[1]))).astype(int)
    
    nToSamplFrom = xTildLongTild_i.shape[0]

    subSamplK=np.ceil(nToSamplFrom/maxSubSamplSiz).astype(int)
    
    if subSamplK==1 : maxSubSamplSiz=nToSamplFrom

    nRuns=np.ceil(subSamplK/3).astype(int)
    
    for k in range(nRuns):
        rowToKeep = np.random.choice(nToSamplFrom, maxSubSamplSiz, replace = False)
        
        x = xTildLongTild_i[rowToKeep, :]
        y = yTildLongTild_i[rowToKeep]
        
        if x.shape[0] > (3 * x.shape[1]):
            Penal.i=runlinear(x=x,y=y, nPredics=nPredics)
            BetaNoInt.k=as((0+(Penal.i$betaNoInt!=0)),"sparseVector")
            EstNoInt.k<-abs(Penal.i$coef_est_noint)
        else:
            Penal.i=runGlmnet(x=x,y=y, nPredics=nPredics, standardize=standardize)
            BetaNoInt.k=as((0+(Penal.i$betaNoInt!=0)),"sparseVector")
            EstNoInt.k<-abs(Penal.i$betaNoInt)
        
        
def runlinear(
        x,
        y,
        nPredics,
        fwerRate=0.25,
        adjust_method="fdr"):
    
    results={}
    nBeta=x.shape[1]
    nObsAll=len(y)
    print("length of y: ",len(y))
    
    #lm_model = LinearRegression(fit_intercept = False)
    #lm_res = lm_model.fit(x, y)
    #lm_res.coef_
    
    lm_res = OLS(y, x).fit()
    p_value_est = lm_res.pvalues
    disc_index = np.arange(0, len(p_value_est), (nPredics+1) )
    p_value_est_noint = np.delete(p_value_est, disc_index, axis=0)
    
    ## this method automatically convert over 1 values to 1
    p_value_est_noint_adj = multipletests(p_value_est_noint, 
                                          alpha = 0.05,
                                          method = adjust_method)[1]

    coef_est = np.abs(lm_res.params)
    disc_index = np.arange(0, len(p_value_est), (nPredics+1) )
    ## NA coef is snot considered here
    coef_est_noint = np.delete(coef_est, disc_index, axis=0)
    
    # return
    results['betaNoInt']=p_value_est_noint_adj<fdrRate
    results['betaInt']=p_value_est
    results['coef_est_noint']=coef_est_noint


    return results
    

  standardize=False
  family="gaussian"
  nfolds=10
  lambda_min_ratio=0.05
  nLam=100
  intercept=True
  zeroSDCut=10**(-20)
  
  

def runGlmnet(
  x,
  y,
  nPredics,
  standardize=False,
  family="gaussian",
  nfolds=10,
  lambda_min_ratio=0.05,
  nLam=100,
  intercept=True,
  zeroSDCut=10**(-20)
):
    results={}
    nBeta=x.shape[1]
    nObsAll=len(y)
    
    # remove near constant x columns
    sdX=np.std(x, axis = 0)
    xWithNearZeroSd=which(sdX<=zeroSDCut)
    if len(xWithNearZeroSd)>0 :
        x = np.delete(x, xWithNearZeroSd, axis=1)
    rm(sdX)
    
    # calculate lambda max
    if family=="gaussian" :
        lamMax = np.max( np.abs( (x*y[:, np.newaxis]).sum(0) ) ) / nObsAll        
        lamVec = np.linspace(lamMax, 0, nLam+1)[0:nLam]  
        cvStartTime = timeit.default_timer()
        
        cvStartTimei = timeit.default_timer()
        
        
        
        cvResul=LassoCV(alphas=lamVec, 
                        fit_intercept=intercept,
                        normalize = standardize,
                        cv = nfolds, 
                        n_jobs = -1).fit(x.copy(), y.copy())
        cvResul.fit(x, y)
        
        cvExeTimei= (timeit.default_timer() - cvStartTimei)/60

        cvResul.alpha
        lamOpi=as.numeric(cvResul$lambda.min)
        cvm=as.vector(cvResul$cvm)*nObsAll
        nLamUsed=length(as.vector(cvResul$lambda))

      rm(cvResul)
    
    

File: IFAA-R-1.0.6/R/originDataScreen.R

#---------------------------------------------------------------------------------------
## Original screen function to select independent reference taxa
#---------------------------------------------------------------------------------------
originDataScreen=function(
  data,
  testCovInd,
  nRef,
  paraJobs,
  refTaxa,
  maxDimensionScr=434*5*10^5,
  sequentialRun,
  allFunc,
  Mprefix,
  covsPrefix,
  binPredInd,
  adjust_method,
  seed){

  results=list()
  
  ## ChangePoint
  np <- reticulate::import("numpy")
  
  # load data info
  basicInfo=dataInfo(data=data,Mprefix=Mprefix,
                     covsPrefix=covsPrefix,
                     binPredInd=binPredInd)

  taxaNames=basicInfo$taxaNames
  nTaxa=basicInfo$nTaxa
  nPredics=basicInfo$nPredics
  gc()

  nNorm=nTaxa-1
  nAlphaNoInt=nPredics*nNorm
  nAlphaSelec=nPredics*nTaxa

  countOfSelec=rep(0,nAlphaSelec)

  # overwrite nRef if the reference taxon is specified
  nRef=length(refTaxa)

  startT1=proc.time()[3]
  if(length(paraJobs)==0){
    availCores=availableCores()
    if(is.numeric(availCores))paraJobs=max(1,availableCores()-2)
    if(!is.numeric(availCores))paraJobs=1
  }
  batch=paraJobs
  forLoopN=ceiling(nRef/batch)

  if(!sequentialRun){
    message(paraJobs, " parallel jobs are registered for analyzing ", nRef, " reference taxa in Phase 1")
  }

  for (jj in 1:forLoopN){

    cl<-parallel::makeCluster(paraJobs)

    parallel::clusterExport(cl=cl, varlist=allFunc,envir=environment())

    doParallel::registerDoParallel(cl)

    if(length(seed)>0){
      set.seed(as.numeric(seed)+jj+10^4)
      parallel::clusterSetRNGStream(cl=cl,(as.numeric(seed)+jj+10^5))
    }

    if(sequentialRun){foreach::registerDoSEQ()}

    #start parallel computing
    if(forLoopN>1 & jj<forLoopN){
      scr1Resu.j=foreach(i=((jj-1)*batch+1):(jj*batch),
                         .multicombine=TRUE,
                         .packages=c("glmnet","Matrix", "reticulate"),
                         .errorhandling="pass") %dopar% {


                           ii=which(taxaNames==refTaxa[i])
                           dataForEst=dataRecovTrans(data=data,ref=refTaxa[i],Mprefix=Mprefix,
                                                     covsPrefix=covsPrefix)

                           xTildLongTild.i=dataForEst$xTildalong
                           yTildLongTild.i=dataForEst$UtildaLong
                           gc()

                           maxSubSamplSiz=min(50000,floor(maxDimensionScr/ncol(xTildLongTild.i)))
                           nToSamplFrom=nrow(xTildLongTild.i)

                             subSamplK=ceiling(nToSamplFrom/maxSubSamplSiz)
                             if(subSamplK==1)maxSubSamplSiz=nToSamplFrom

                             nRuns=ceiling(subSamplK/3)
                             # nRuns = 10 # ChangePoint
                             # maxSubSamplSiz = nToSamplFrom / 2
                             # RunLinear / RunGlmnet
                             # CI INFERENCE  RunLinear / RunGlmnet
                             
                             # ====== #
                             
                             ## Zero SD variable
                             ## v6 = all 0
                             ## v7 = all 0
                             
                             ## ChangePoint
                             nRuns = 3 
                             maxSubSamplSiz = nToSamplFrom / 2
                             
                             # if (ii == 17) {
                             #   browser()
                             # }
                             
                             for (k in 1:nRuns){
                               # rowToKeep=sample(nToSamplFrom,maxSubSamplSiz)
                               # x=as((xTildLongTild.i[rowToKeep,]),"sparseMatrix")
                               # y=yTildLongTild.i[rowToKeep]
                               
                               ## ChangePoint
                               np$random$seed(as.integer(seed+k-1))
                               rowToKeep=np$random$choice(as.integer(nToSamplFrom), as.integer(maxSubSamplSiz), replace=FALSE)+1
                               rowToKeep <- as.integer(rowToKeep)
                               x=as((xTildLongTild.i[rowToKeep,]),"sparseMatrix")
                               y=yTildLongTild.i[rowToKeep]


                               if (F){ # (dim(x)[1]>(3*dim(x)[2])) {
                                 Penal.i=runlinear(x=x,y=y,
                                                   nPredics=nPredics)
                                 BetaNoInt.k=as((0+(Penal.i$betaNoInt!=0)),"sparseVector")
                                 EstNoInt.k<-abs(Penal.i$coef_est_noint)
                               } else {
                                 Penal.i=runGlmnet(x=x,y=y,
                                                   nPredics=nPredics)
                                 BetaNoInt.k=as((0+(Penal.i$betaNoInt!=0)),"sparseVector")
                                 EstNoInt.k<-abs(Penal.i$betaNoInt)
                               }
                               if(k==1) {
                                 BetaNoInt.i=BetaNoInt.k
                                 EstNoInt.i=EstNoInt.k
                               }
                               if(k>1) {
                                 BetaNoInt.i=BetaNoInt.i+BetaNoInt.k
                                 EstNoInt.i=EstNoInt.i+EstNoInt.k
                               }
                             }

                             # if (ii == 17) {
                             #   browser()
                             # }
                             BetaNoInt.i=BetaNoInt.i/nRuns
                             EstNoInt.i=EstNoInt.i/nRuns

                           selection.i=as(rep(0,nAlphaSelec),"sparseVector")
                           coef.i<-rep(0,nAlphaSelec)
                           if (ii==1){
                             selection.i[-seq(1,nPredics)]=BetaNoInt.i
                             coef.i[-seq(1,nPredics)]=EstNoInt.i
                           }
                           if (ii==nTaxa) {
                             selection.i[-seq((nAlphaSelec-nPredics+1),nAlphaSelec)]=BetaNoInt.i
                             coef.i[-seq((nAlphaSelec-nPredics+1),nAlphaSelec)]=EstNoInt.i
                           }
                           if ((ii>1) & (ii<nTaxa)) {
                             selection.i[1:(nPredics*(ii-1))]=BetaNoInt.i[1:(nPredics*(ii-1))]
                             selection.i[(nPredics*ii+1):nAlphaSelec]=BetaNoInt.i[(nPredics*(ii-1)+1):nAlphaNoInt]
                             coef.i[1:(nPredics*(ii-1))]=EstNoInt.i[1:(nPredics*(ii-1))]
                             coef.i[(nPredics*ii+1):nAlphaSelec]=EstNoInt.i[(nPredics*(ii-1)+1):nAlphaNoInt]
                           }
                           # create return vector
                           recturnlist=list()
                           recturnlist[[1]]=selection.i
                           recturnlist[[2]]<-coef.i
                           return(recturnlist)
                         }
      parallel::stopCluster(cl)
      gc()
      if(jj==1)scr1Resu=scr1Resu.j
      if(jj>1)scr1Resu=do.call(base::c,list(scr1Resu,scr1Resu.j))
    }

    if(jj==forLoopN){
      scr1Resu.j=foreach(i=((forLoopN-1)*batch+1):nRef,
                         .multicombine=TRUE,
                         .packages=c("glmnet","Matrix", "reticulate"),
                         .errorhandling="pass") %dopar% {


                           ii=which(taxaNames==refTaxa[i])
                           dataForEst=dataRecovTrans(data=data,ref=refTaxa[i],Mprefix=Mprefix,
                                                     covsPrefix=covsPrefix)

                           xTildLongTild.i=dataForEst$xTildalong
                           yTildLongTild.i=dataForEst$UtildaLong
                           gc()

                           maxSubSamplSiz=min(50000,floor(maxDimensionScr/ncol(xTildLongTild.i)))
                           nToSamplFrom=nrow(xTildLongTild.i)

                             subSamplK=ceiling(nToSamplFrom/maxSubSamplSiz)
                             if(subSamplK==1)maxSubSamplSiz=nToSamplFrom

                             nRuns=ceiling(subSamplK/3)

                             ## ChangePoint
                             nRuns = 3 
                             maxSubSamplSiz = nToSamplFrom / 2
                             

                             
                             for (k in 1:nRuns){
                               # rowToKeep=sample(nToSamplFrom,maxSubSamplSiz)
                               # x=as((xTildLongTild.i[rowToKeep,]),"sparseMatrix")
                               # y=(yTildLongTild.i[rowToKeep])
                               
                               ## ChangePoint
                               np$random$seed(as.integer(seed+k-1))
                               rowToKeep=np$random$choice(as.integer(nToSamplFrom), as.integer(maxSubSamplSiz), replace=FALSE)+1
                               rowToKeep <- as.integer(rowToKeep)
                               x=as((xTildLongTild.i[rowToKeep,]),"sparseMatrix")
                               y=yTildLongTild.i[rowToKeep]

                               if (F){ #(dim(x)[1]>(3*dim(x)[2])) {
                                 Penal.i=runlinear(x=x,y=y,
                                                   nPredics=nPredics)
                                 BetaNoInt.k=as((0+(Penal.i$betaNoInt!=0)),"sparseVector")
                                 EstNoInt.k<-abs(Penal.i$coef_est_noint)
                               } else {
                                 Penal.i=runGlmnet(x=x,y=y,
                                                   nPredics=nPredics)
                                 BetaNoInt.k=as((0+(Penal.i$betaNoInt!=0)),"sparseVector")
                                 EstNoInt.k<-abs(Penal.i$betaNoInt)
                               }

                               gc()
                               if(k==1) {
                                 BetaNoInt.i=BetaNoInt.k
                                 EstNoInt.i=EstNoInt.k
                               }
                               if(k>1) {
                                 BetaNoInt.i=BetaNoInt.i+BetaNoInt.k
                                 EstNoInt.i=EstNoInt.i+EstNoInt.k
                               }
                             }
                             gc()
                             BetaNoInt.i=BetaNoInt.i/nRuns
                             EstNoInt.i=EstNoInt.i/nRuns

                           selection.i=as(rep(0,nAlphaSelec),"sparseVector")
                           coef.i<-rep(0,nAlphaSelec)
                           if (ii==1){
                             selection.i[-seq(1,nPredics)]=BetaNoInt.i
                             coef.i[-seq(1,nPredics)]=EstNoInt.i
                           }
                           if (ii==nTaxa) {
                             selection.i[-seq((nAlphaSelec-nPredics+1),nAlphaSelec)]=BetaNoInt.i
                             coef.i[-seq((nAlphaSelec-nPredics+1),nAlphaSelec)]=EstNoInt.i
                           }
                           if ((ii>1) & (ii<nTaxa)) {
                             selection.i[1:(nPredics*(ii-1))]=BetaNoInt.i[1:(nPredics*(ii-1))]
                             selection.i[(nPredics*ii+1):nAlphaSelec]=BetaNoInt.i[(nPredics*(ii-1)+1):nAlphaNoInt]
                             coef.i[1:(nPredics*(ii-1))]=EstNoInt.i[1:(nPredics*(ii-1))]
                             coef.i[(nPredics*ii+1):nAlphaSelec]=EstNoInt.i[(nPredics*(ii-1)+1):nAlphaNoInt]
                           }
                           # create return vector
                           recturnlist=list()
                           recturnlist[[1]]=selection.i
                           recturnlist[[2]]<-coef.i
                           return(recturnlist)
                         }
      parallel::stopCluster(cl)
      if(forLoopN==1)scr1Resu=scr1Resu.j
      if(forLoopN>1)scr1Resu=do.call(base::c,list(scr1Resu,scr1Resu.j))
      gc()
    }

    # if(jj>0 & (jj%%(ceiling(forLoopN/10))==0)){
    #   message(round(100*jj/forLoopN,0), " percent of phase 1 analysis has been done")
    # }
  }

  endT=proc.time()[3]


  selecList=list()
  for(i in 1:nRef){
    selecList[[i]]=scr1Resu[[i]][[1]]
  }




  estList=list()
  for(i in 1:nRef){
    estList[[i]]=scr1Resu[[i]][[2]]
  }

  selecList<- lapply(selecList, as, "sparseMatrix")
  estList<-lapply(estList,as.matrix)
  scr1ResuSelec=do.call(cbind, selecList)
  scr1ResuEst<-do.call(cbind,estList)




  # create count of selection for individual testCov
  countOfSelecForAllPred=as(matrix(Matrix::rowSums(scr1ResuSelec),nrow=nPredics),"sparseMatrix")
  EstOfAllPred<-matrix(rowSums(scr1ResuEst),nrow=nPredics)

  testCovCountMat=countOfSelecForAllPred[testCovInd,,drop=FALSE]
  testEstMat<-EstOfAllPred[testCovInd,,drop=FALSE]

  # create overall count of selection for all testCov as a whole
  countOfSelecForAPred=as(matrix(Matrix::colSums(testCovCountMat),nrow=1),"sparseMatrix")
  estOfSelectForAPred<-matrix(Matrix::colSums(testEstMat),nrow=1)
  gc()

  colnames(countOfSelecForAPred)=taxaNames
  colnames(estOfSelectForAPred)<-taxaNames

  # return results
  results$testCovCountMat=testCovCountMat
  results$testEstMat<-testEstMat
  results$countOfSelecForAPred=countOfSelecForAPred
  results$estOfSelectForAPred<-estOfSelectForAPred
  return(results)
}
